catalog/rocks/pgiterator.go

```diff
@@ -0,0 +1,288 @@
+package rocks
+
+import (
+	"context"
+
+	"github.com/treeverse/lakefs/db"
+)
+
+const (
+	// IteratorPrefetchSize is the amount of records to prefetch from PG
+	IteratorPrefetchSize = 1000
+)
+
+type pgBranchRecord struct {
+	BranchID `db:"id"`
+	*PgBranch
+}
+
+type PGRepositoryIterator struct {
+	db  db.Database
+	ctx context.Context
+
+	value *RepositoryRecord
+	buf   []*RepositoryRecord
+
+	offset           string
+	prefetchSize     int
+	prefetchRequired bool
+
+	err error
+}
+
+func NewRepositoryIterator(ctx context.Context, db db.Database, prefetchSize int, offset string) *PGRepositoryIterator {
+	return &PGRepositoryIterator{
+		db:               db,
+		ctx:              ctx,
+		prefetchSize:     prefetchSize,
+		prefetchRequired: true,
+		offset:           offset,
+	}
+}
+
+func (ri *PGRepositoryIterator) Next() bool {
+	// no buffer is initialized
+	if ri.buf == nil || len(ri.buf) == 0 {
+		ri.prefetch()
+	}
+
+	if len(ri.buf) == 0 {
+		return false
+	}
+
+	// stage a value and increment offset
+	ri.value = ri.buf[0]
+	ri.offset = string(ri.value.RepositoryID)
+	if len(ri.buf) >= 2 {
+		ri.buf = ri.buf[1:]
+	} else {
+		ri.buf = make([]*RepositoryRecord, 0)
+	}
+
+	return true
+
+}
+
+func (ri *PGRepositoryIterator) prefetch() bool {
+	if !ri.prefetchRequired {
+		return false
+	}
+	err := ri.db.WithContext(ri.ctx).Select(&ri.buf, `
+			SELECT id, storage_namespace, creation_date, default_branch
+			FROM kv_repositories
+			WHERE id > $1
+			ORDER BY id ASC
+			LIMIT $2`, ri.offset, ri.prefetchSize)
+	if err != nil {
+		ri.err = err
+		return false
+	}
+	if len(ri.buf) < ri.prefetchSize {
+		ri.prefetchRequired = false
+	}
+	return len(ri.buf) > 0
+}
+
+func (ri *PGRepositoryIterator) SeekGE(id RepositoryID) bool {
+	ri.offset = string(id)
+	ri.prefetchRequired = true
+	ri.buf = make([]*RepositoryRecord, 0)
+	ri.prefetch()
```

Comment:
`offset` is `id`, and `prefetch` says `id > $1`, so AFAICT this function is really `SeekGT`.  (Which admittedly makes more sense for most iteration scenarios, unless you want to call `SeekGE(after + "\0")`).

Reply:
yep, that's a bug. I'll add a test. prefetch should do `id >= $1`